### PR TITLE
Fix Snapshots Recording Incorrect Max. Segment Counts

### LINKED ISSUE #74249: Max Segment Count in Snapshot Index Details from Get Snapshots API can be Wrong
There appears to be a bug in the logic that computes the segment count of a shard snapshot in some background merge scenarios.

The segment count is taken from the index commit that we are snapshotting via a call to `org.apache.lucene.index.IndexCommit#getSegmentCount`. This is correct for all snapshots except for those that use the sequence number based shortcut introduced in https://github.com/elastic/elasticsearch/pull/52182 that will prevent an updated set of files from getting uploaded when there already is a shard copy at the same sequence number that is in the current index commit.
This means that if a snapshot is taken once before a background merge and once after without any indexing into the shard, then the segment count stored for the second snapshot will be wrong if segment counts for the shard changed. We can fix this by taking the segment count from the files list instead in this case.

### PR DESCRIPTION
**Pull request by original-brownbear**
If sequence numbers are equal across snapshots (and thus no files get written to the repository)
the segment count in the index commit is not reliable as it may have changed due to background merges.
=> fixed by always using the segment count determined from the file names in the snapshot instead

closes #74249

I'm not the biggest fan of my test here to be honest, but I couldn't find a way to have more fine grained control over background merges  so I went with this rather brute force approach (alternative test suggestions very welcome :)). The test does however reproduce the problem reliable in 100% of runs for me.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Could we set `index.merge.enabled: false` to block merges for the first snapshot, and then remove that setting (close & reopen the index) for the second snapshot. Might need to force-merge after reopening to trigger a merge, but that won't update the force-merge ID if `maxNumSegments` ≤ 0.

**Comment by original-brownbear** [comment]
> Could we set index.merge.enabled: false to block merges for the first snapshot,

Done, so much cuter :) Thanks!

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/BlobStoreIncrementalityIT.java [hunk 1]
```diff
@@ -166,6 +183,74 @@ public void testForceMergeCausesFullSnapshot() throws Exception {
...
+
+        // create an empty snapshot so that later snapshots run as quickly as possible
+        createFullSnapshot(repoName, "empty");
+
+        // create a situation where we temporarily have a bunch of segments until the merges can catch up
+        long id = 0;
+        final int rounds = scaledRandomIntBetween(50, 300);
+        for (int i = 0; i < rounds; ++i) {
+            final int numDocs = scaledRandomIntBetween(100, 1000);
+            BulkRequestBuilder request = client().prepareBulk();
```
Comment (DaveCTurner) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/BlobStoreIncrementalityIT.java:202]: I think it'd be better to refresh on each bulk to force a segment to be written:

```suggestion
            BulkRequestBuilder request = client().prepareBulk().setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
```

I suspect if we did that we wouldn't need so many rounds.

Comment (original-brownbear) [reply server/src/internalClusterTest/java/org/elasticsearch/snapshots/BlobStoreIncrementalityIT.java]: ++ right done in d629c74

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/BlobStoreIncrementalityIT.java [hunk 2]
```diff
@@ -166,6 +184,73 @@ public void testForceMergeCausesFullSnapshot() throws Exception {
...
+        final long beforeSegmentCount = beforeIndexDetails.getMaxSegmentsPerShard();
+
+        // reactivate merges
+        assertAcked(admin().indices().prepareClose(indexName).get());
+        assertAcked(
+            admin().indices()
+                .prepareUpdateSettings(indexName)
+                .setSettings(
+                    Settings.builder()
+                        .put(MergePolicyConfig.INDEX_MERGE_POLICY_SEGMENTS_PER_TIER_SETTING.getKey(), "2")
```
Comment (DaveCTurner) [inline server/src/internalClusterTest/java/org/elasticsearch/snapshots/BlobStoreIncrementalityIT.java:226]: 👍  we know we did ≥3 rounds, each of which made ≥1 segment, and they're all pretty small (≤2MB) so this guarantees a merge IIUC.

**Review by DaveCTurner — APPROVED** [review]
LGTM

File: server/src/internalClusterTest/java/org/elasticsearch/snapshots/BlobStoreIncrementalityIT.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/internalClusterTest/java/org/elasticsearch/snapshots/BlobStoreIncrementalityIT.java:226]: ++ that was assumption here. Also ran a few hundred rounds of this to verify and it seems to hold :crossed_fingers:

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**